File: src/stores/contacts.ts
/**
 * Contacts store - manages contact list.
 */

import { create } from 'zustand';
import { invoke } from '@tauri-apps/api/core';
import type { Contact, TrustLevel } from '../types';

interface ContactsStore {
  contacts: Contact[];
  isLoading: boolean;
  error: string | null;
  
  // Actions
  loadContacts: () => Promise<void>;
  addContact: (peerId: string, alias?: string, publicKey?: number[]) => Promise<Contact>;
  addContactFromQR: (qrData: string, alias?: string) => Promise<Contact>;
  updateAlias: (peerId: string, alias: string) => Promise<void>;
  updateTrust: (peerId: string, trustLevel: TrustLevel) => Promise<void>;
  deleteContact: (peerId: string) => Promise<void>;
  getContact: (peerId: string) => Contact | undefined;
  clearError: () => void;
}

export const useContactsStore = create<ContactsStore>((set, get) => ({
  contacts: [],
  isLoading: false,
  error: null,

  loadContacts: async () => {
    set({ isLoading: true, error: null });
    try {
      const result = await invoke<Array<{
        peer_id: string;
        alias: string;
        public_key: number[];
        trust_level: string;
        created_at: number;
        updated_at: number;
      }>>('get_contacts');
      
      const contacts: Contact[] = result.map(c => ({
        peerId: c.peer_id,
        alias: c.alias,
        publicKey: Array.from(c.public_key).map(b => b.toString(16).padStart(2, '0')).join(''),
        trustLevel: c.trust_level as TrustLevel,
        createdAt: c.created_at,
        updatedAt: c.updated_at,
      }));
      
      set({ contacts, isLoading: false });
    } catch (e) {
      set({ error: String(e), isLoading: false });
    }
  },

  addContact: async (peerId: string, alias?: string, publicKey?: number[]) => {
    set({ isLoading: true, error: null });
    try {
      const result = await invoke<{
        peer_id: string;
        alias: string;
        public_key: number[];
        trust_level: string;
        created_at: number;
        updated_at: number;
      }>('add_contact', { peerId, alias, publicKey: publicKey ?? [] });
      
      const contact: Contact = {
        peerId: result.peer_id,
        alias: result.alias,
        publicKey: Array.from(result.public_key).map(b => b.toString(16).padStart(2, '0')).join(''),
        trustLevel: result.trust_level as TrustLevel,
        createdAt: result.created_at,
        updatedAt: result.updated_at,
      };
      
      set(state => ({
        contacts: [...state.contacts, contact],
        isLoading: false,
      }));
      
      return contact;
    } catch (e) {
      set({ error: String(e), isLoading: false });
      throw e;
    }
  },

  addContactFromQR: async (qrData: string, alias?: string) => {
    set({ isLoading: true, error: null });
    try {
      const result = await invoke<{
        peer_id: string;
        alias: string;
        public_key: number[];
        trust_level: string;
        created_at: number;
        updated_at: number;
      }>('add_contact_from_qr', { qrData, alias });
      
      const contact: Contact = {
        peerId: result.peer_id,
        alias: result.alias,
        publicKey: Array.from(result.public_key).map(b => b.toString(16).padStart(2, '0')).join(''),
        trustLevel: result.trust_level as TrustLevel,
        createdAt: result.created_at,
        updatedAt: result.updated_at,
      };
      
      set(state => ({
        contacts: [...state.contacts, contact],
        isLoading: false,
      }));
      
      return contact;
    } catch (e) {
      set({ error: String(e), isLoading: false });
      throw e;
    }
  },

  updateAlias: async (peerId: string, alias: string) => {
    try {
      await invoke('update_contact_alias', { peerId, alias });
      set(state => ({
        contacts: state.contacts.map(c =>
          c.peerId === peerId ? { ...c, alias, updatedAt: Date.now() / 1000 } : c
        ),
      }));
    } catch (e) {
      set({ error: String(e) });
    }
  },

  updateTrust: async (peerId: string, trustLevel: TrustLevel) => {
    try {
      await invoke('update_contact_trust', { peerId, trustLevel });
      set(state => ({
        contacts: state.contacts.map(c =>
          c.peerId === peerId ? { ...c, trustLevel, updatedAt: Date.now() / 1000 } : c
        ),
      }));
    } catch (e) {
      set({ error: String(e) });
    }
  },

  deleteContact: async (peerId: string) => {
    try {
      await invoke('delete_contact', { peerId });
      set(state => ({
        contacts: state.contacts.filter(c => c.peerId !== peerId),
      }));
    } catch (e) {
      set({ error: String(e) });
    }
  },

  getContact: (peerId: string) => {
    return get().contacts.find(c => c.peerId === peerId);
  },

  clearError: () => set({ error: null }),
}));

File: src-tauri/src/commands/contacts.rs
//! Contact commands.

use crate::commands::AppState;
use crate::storage::contacts::{Contact, TrustLevel};
use serde::Serialize;
use tauri::State;

/// Error wrapper for commands
#[derive(Debug, Serialize)]
pub struct CommandError {
    pub message: String,
}

/// Get all contacts.
#[tauri::command]
pub async fn get_contacts(
    state: State<'_, AppState>,
) -> Result<Vec<Contact>, CommandError> {
    let db = state.db.lock().await;
    let db = db.as_ref().ok_or_else(|| CommandError { message: "Database not initialized".into() })?;
    
    crate::storage::contacts::get_contacts(db)
        .await
        .map_err(|e| CommandError { message: e.to_string() })
}

/// Add a new contact.
#[tauri::command]
pub async fn add_contact(
    state: State<'_, AppState>,
    peer_id: String,
    alias: Option<String>,
    public_key: Vec<u8>,
) -> Result<Contact, CommandError> {
    let db = state.db.lock().await;
    let db = db.as_ref().ok_or_else(|| CommandError { message: "Database not initialized".into() })?;
    
    let now = chrono::Utc::now().timestamp();
    let contact = Contact {
        peer_id: peer_id.clone(),
        alias: alias.unwrap_or_else(|| peer_id[..8.min(peer_id.len())].to_string()),
        public_key,
        trust_level: TrustLevel::Unknown,
        created_at: now,
        updated_at: now,
    };
    
    crate::storage::contacts::add_contact(db, &contact)
        .await
        .map_err(|e| CommandError { message: e.to_string() })?;
    
    Ok(contact)
}

/// Add a contact from QR code data.
#[tauri::command]
pub async fn add_contact_from_qr(
    state: State<'_, AppState>,
    qr_data: String,
    alias: Option<String>,
) -> Result<Contact, CommandError> {
    // Parse: whisper://<peer_id>/<public_key_hex>
    let parts: Vec<&str> = qr_data
        .strip_prefix("whisper://")
        .ok_or_else(|| CommandError { message: "Invalid QR format".into() })?
        .split('/')
        .collect();
    
    if parts.len() != 2 {
        return Err(CommandError { message: "Invalid QR format".into() });
    }
    
    let peer_id = parts[0].to_string();
    let public_key = hex::decode(parts[1])
        .map_err(|e| CommandError { message: format!("Invalid public key: {}", e) })?;
    
    add_contact(state, peer_id, alias, public_key).await
}

/// Update contact alias.
#[tauri::command]
pub async fn update_contact_alias(
    state: State<'_, AppState>,
    peer_id: String,
    alias: String,
) -> Result<(), CommandError> {
    let db = state.db.lock().await;
    let db = db.as_ref().ok_or_else(|| CommandError { message: "Database not initialized".into() })?;
    
    let mut contact = crate::storage::contacts::get_contact(db, &peer_id)
        .await
        .map_err(|e| CommandError { message: e.to_string() })?
        .ok_or_else(|| CommandError { message: "Contact not found".into() })?;
    
    contact.alias = alias;
    contact.updated_at = chrono::Utc::now().timestamp();
    
    crate::storage::contacts::update_contact(db, &contact)
        .await
        .map_err(|e| CommandError { message: e.to_string() })
}

/// Update contact trust level.
#[tauri::command]
pub async fn update_contact_trust(
    state: State<'_, AppState>,
    peer_id: String,
    trust_level: String,
) -> Result<(), CommandError> {
    let db = state.db.lock().await;
    let db = db.as_ref().ok_or_else(|| CommandError { message: "Database not initialized".into() })?;
    
    let mut contact = crate::storage::contacts::get_contact(db, &peer_id)
        .await
        .map_err(|e| CommandError { message: e.to_string() })?
        .ok_or_else(|| CommandError { message: "Contact not found".into() })?;
    
    contact.trust_level = match trust_level.as_str() {
        "verified" => TrustLevel::Verified,
        "unverified" => TrustLevel::Unverified,
        _ => TrustLevel::Unknown,
    };
    contact.updated_at = chrono::Utc::now().timestamp();
    
    crate::storage::contacts::update_contact(db, &contact)
        .await
        .map_err(|e| CommandError { message: e.to_string() })
}

/// Delete a contact.
#[tauri::command]
pub async fn delete_contact(
    state: State<'_, AppState>,
    peer_id: String,
) -> Result<(), CommandError> {
    let db = state.db.lock().await;
    let db = db.as_ref().ok_or_else(|| CommandError { message: "Database not initialized".into() })?;
    
    crate::storage::contacts::delete_contact(db, &peer_id)
        .await
        .map_err(|e| CommandError { message: e.to_string() })
}
